src/lib/storage.ts

```diff
@@ -0,0 +1,34 @@
+import {type StorageProvider, type SyncConfig} from "@welshman/store"
+import {Preferences} from "@capacitor/preferences"
+
+export class PreferencesStorageProvider implements StorageProvider {
+  get = async (key: string): Promise<any> => {
+    const result = await Preferences.get({key})
+    if (!result.value) return undefined
+    try {
+      return JSON.parse(result.value)
+    } catch (e) {
+      return undefined
+    }
+  }
+
+  set = async (key: string, value: any): Promise<void> => {
+    await Preferences.set({key, value: JSON.stringify(value)})
+  }
+}
+
+// singleton instance of PreferencesStorageProvider
+export const preferencesStorageProvider = new PreferencesStorageProvider()
+
+// Temporary re-implementation of sync function to fix race condition
+export const sync = async <T>({key, store, storage}: SyncConfig<T>) => {
+  const storedValue = await storage.get(key)
+
+  if (storedValue !== undefined) {
+    store.set(storedValue)
+  }
+
+  store.subscribe(async (value: T) => {
+    await storage.set(key, value)
+  })
```

Comment:
The await here isn't going to do anything, if `subscribe` gets triggered many times in quick succession this will result in concurrent updates. If `set` doesn't consistently take the same time, older updates might overwrite newer ones, depending on the `storage` implementation. I suppose we could move the concurrency management requirement down to the storage implementation; it doesn't look like `preferences` has a way to abort `set` calls, so it's probably ok, but if there is the possibility of a race condition we should stack promises to ensure ordered execution. So in `storage.set`:

```
p = Promise.resolve()
set = async (key: string, value: T) => {
  await this.p
  this.p = Preferences.set({key, value: JSON.stringify(value)})
  await this.p
}
```

Reply:
Also, go ahead and open a PR to welshman with this change once it's sorted out.